### ISSUE TITLE
Possible to overlay html elements on top of MKMapView?

### ISSUE THREAD

**Issue by iori57** [victorzimmer/MapKit#3]
Firstly, thanks for the plugin it works great!

Is it possible to put the MKMapView under the webview, such as how cordova-plugin-googlemaps (https://github.com/mapsplugin/cordova-plugin-googlemaps) do it? most part of the webview will be transparent to allow the MKMapView to be displayed but this way we can overlay html elements such as menus if we wanted to.

Not sure if this is possible in iOS, found several cordova plugins for MapKit but so far none are doing this, there must be a reason or simply impossible?

Thanks!

**Comment by victorzimmer** [comment]
Happy to hear that it works for you!

Already working on it. From what I can see it should be possible, I currently have transparent webViews working. It will be in version 0.0.4.

The webView will of course be transparent and you’ll be able to choose wether or not it should have UserInteraction (If it does the map underneath cant respond to touch).

—
Victor Zimmer

**Comment by iori57** [comment]
Thanks for the fast response!

That is great news.. looking forward to it :)

If you have a development branch I can help to test it too before the actual release. (the truth is I am too excited to try it out XD)

Thanks again!

**Comment by victorzimmer** [comment]
Well I should probably use branches (Everything so far has been committed directly to Master :P). It should probably be ready tonight or tomorrow so I'll notify you by posting here as soon as it's committed.

Also I'm gonna make a development branch right away.

**Comment by iori57** [comment]
Awesome, thanks!

**Comment by victorzimmer** [comment]
No problem!

**Comment by victorzimmer** [comment]
![simulator screen shot 19 nov 2015 03 17 24](https://cloud.githubusercontent.com/assets/715456/11260503/37166336-8e6c-11e5-88ba-d90e1895b585.png)

Thought you might like this ;)

**Comment by iori57** [comment]
Neat stuff!

**Comment by victorzimmer** [comment]
The version in the development branch will now make a web view with every map view. It will look in your "www" folder for a file named with the ID you gave your map. 

So if you create the map as `MKInterface.MKMap('supercool')` it will look in `www` for `supercool.html`.

**Comment by iori57** [comment]
Anyway to make it work with existing webview? 

For example cordova loads my index.html. Pressing one button will then display the mapkit mapview behind it.

Will try out the development branch when I am able :)

**Comment by victorzimmer** [comment]
As far as I can see there is no good way to modify the existing Cordova webView without breaking things. However I am working on communication between the MapKit WebView and the Cordova WebView so you'll be able to click things in the overlay and have the app react to the event.

**Comment by iori57** [comment]
I have initial success in trying to put the mapview behind of existing cordova webview if you are interested.

I followed how https://github.com/mapsplugin/cordova-plugin-googlemaps did it with Android + GoogleMap, just apply the same concept to iOS + MapKit

Steps:
- In MapKit.m createMapView(), instead of adding the mapview as a subview to cordova webview, add it as a sibling instead, as such

> [webView.superview insertSubview:mapView belowSubview:webView];
> [webView setBackgroundColor:[UIColor clearColor]];
- In MainViewController.m of cordova, uncomment the block of code for webViewDidStartLoad() and add the below:

> [self.webView setOpaque:NO];

For some reason this line does not work if I put it in createMapView(), based on some answers in stackoverflow there are comments that it will only work if we set it earlier, right after we create the webview.
- In the HTML side of things, set the div that suppose to show the mapview, along with all its parent divs, to have background-color: rgba(0, 0, 0, 0) !important;

With this the native map can be shown in the app, while opening menu/side panel and stuff can be properly overlay on top of it.

What is missing is that we just need to add two more plugin methods, to disable/enable webview interaction depending whether we want to interact with the webview (menu/side panel) or the map itself.

**Comment by victorzimmer** [comment]
Well that might work. I do still believe that a separate view is a better approach though. Especially since that view will appear and disappear with the map. All I need to do is to implement a way for the views to message each other.

—
Victor Zimmer

**Comment by victorzimmer** [comment]
Join me on Freenode IRC? In the #cordova channel

—
Victor Zimmer

**Comment by iori57** [comment]
Sorry I missed this, are you still in the channel?

**Comment by victorzimmer** [comment]
Yeah

—
Victor Zimmer

**Comment by iori57** [comment]
In the screenshot with google search html overlay, are you able to interact with both the search box and the native map in the background at the same time?

**Comment by victorzimmer** [comment]
No. Sadly it seems you'll have to make a choice :(

—
Victor Zimmer

**Comment by iori57** [comment]
I found a way to do it, https://github.com/sSegev/SSAlphaPass

First subclass the UIWebView:

```
@interface AlphaPassUIWebView : UIWebView
- (UIColor *) colorOfPoint:(CGPoint)point withLayer:(CALayer*)layer;
@property(nonatomic,getter=isAlphaPassEnabled) BOOL alphaPassEnabled;
@end
```

Implement colorOfPoint function to detect transparent pixels and override pointInside function and make it such that transparent pixels can be clicked through, while non transparent pixels will consume the click:

```
@implementation AlphaPassUIWebView

- (UIColor *) colorOfPoint:(CGPoint)point withLayer:(CALayer*)layer
{
    unsigned char pixel[4] = {0};

    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();

    CGContextRef context = CGBitmapContextCreate(pixel, 1, 1, 8, 4, colorSpace, kCGBitmapAlphaInfoMask & kCGImageAlphaPremultipliedLast);

    CGContextTranslateCTM(context, -point.x, -point.y);

    [layer renderInContext:context];

    CGContextRelease(context);
    CGColorSpaceRelease(colorSpace);

    UIColor *color = [UIColor colorWithRed:pixel[0]/255.0 green:pixel[1]/255.0 blue:pixel[2]/255.0 alpha:pixel[3]/255.0];

    return color;
}

-(BOOL)pointInside:(CGPoint)point withEvent:(UIEvent *)event
{
    if (_alphaPassEnabled)
    {
        UIColor *pixelColor = [self colorOfPoint:point withLayer:self.layer];
        if(!CGColorGetAlpha(pixelColor.CGColor))
        {
            return NO;
        }
    }

    return YES;
}

@end
```

When you init the webview, just use the subclass instead:

```
AlphaPassUIWebView* webView = [[AlphaPassUIWebView alloc] initWithFrame:CGRectMake(xPos, yPos, width, height)];
```

And set the flag if you need to make the transparent area of the webview to 'click through' to the map view behind. 

```
[webView setAlphaPassEnabled:YES];
```

This way you can interact with the webview html elements as well as the map view behind it :)

**Comment by victorzimmer** [comment]
That is just extremely neat, have you tested it?

**Comment by victorzimmer** [comment]
I'll see if I can get to implement it in MapKit tomorrow

**Comment by iori57** [comment]
Yup tested and working for my case. I overrode the default cordova webview to use this subclass, put the mapview behind the webview and use it just like I described a few post before this :)

**Comment by seanislegend** [comment]
This sounds like a good solution, do you have this example in a repo somewhere, @iori57 ? Sorry, I'm not a Obj-C person so stuck on this one! 

Thanks to you and @victorzimmer for the great work on this!

**Comment by victorzimmer** [comment]
@seanislegend I have a working demo, but I'm still trying to figure out how to let JavaScript communicate between the WebViews. My current solution is having ObjC relay it and use exec().

**Comment by victorzimmer** [comment]
That is of course a version where the WebView is on top of the MapView, with the background being transparent and relaying touch events to the Map.

**Comment by seanislegend** [comment]
Okay, cool. I found this repo after having switched from the Google Maps plugin and not being able to have elements showing on top of the map has rendered quite a bit of our interface null and void - not your fault, of course!

If there's anything I can contribute to I'm happy to help. Thanks again.

**Comment by victorzimmer** [comment]
Sorry to hear that, I currently have some features I'm working on, including this and support for all platforms. I'll see if I can't finish a version that lets you do this in the next few days.

**Comment by seanislegend** [comment]
No need to be sorry! Not complaining. Merely offering feedback and asking if I can help. More than happy to test also so do let me know if I can be of assistance.

**Comment by petercn** [comment]
Any progress on this guys? I could really use a repo as an example of it myself too, I'm not a big objective c guy either :-/

**Comment by victorzimmer** [comment]
Sorry to say I haven't got time to look at this yet, but I'll see if I can't get some time soon. Still trying to figure out a better way for JS to communicate between the two WebViews, but might just go with exec's and create a little JS wrapper.

**Comment by victorzimmer** [comment]
So I just got to take a look at it. Current state is that the MapKit code is a mess (my fault). Mostly because this was just a little project I worked on for myself and never really meant to do everything it tries to do now.

In the coming days, provided I get the time I'll try to clean up the code or potentially recreating MapKit with a cleaner structure in a new project. 

This is severely needed for advanced things like HTML elements over the Map which currently just doesn't fit the structure MapKit has. It will also be an enormous benefit for me when doing cross-platform support and supporting new Cordova versions. 

TLDR; I'm extremely sorry for the time this is taking, and thank you for your patience. This might be my discontinuation of MapKit for the sake of making something better in a new project. More info to come as I get to take a look at it tomorrow.

-Victor

**Comment by seanislegend** [comment]
Thanks for your time on this Victor. It sounds like your project has been hijacked by other people, but hopefully in a good way! If there's anything I can contribute JavaScript wise please let me know, I am happy to help where I can. 

Thanks again.

**Comment by victorzimmer** [comment]
Will do indeed. With my current approach to this overlay issue there will be a bunch of JS to "wrap".

**Comment by petercn** [comment]
Hi guys, just checking in again to see if there's been any progress or if there's at least a bare bones example out there of a native map underneath a transparent webview that allows the user to move the native map below when they touch upon any transparent area of the webview up top?

**Comment by victorzimmer** [comment]
Hello Peter, barely got to start creating MapKit 2 before my focus was drawn away by real life and moving to an entirely new country.

It is definitely possible to do this and I remember having a semi-functional version, but I was struggling to figure out how JS would communicate between the two.
